### DISCUSSION TITLE
Calibration Substrate in Reference Format

### DISCUSSION THREAD

**Discussion by Zwelckovich** [scikit-rf/scikit-rf#880] [Q&A] [answered]
Hello,
to create calkits iam using the code from the documentation [link](https://scikit-rf.readthedocs.io/en/latest/examples/metrology/SOLT%20Calibration%20Standards%20Creation.html).
There described formats are R&S and Keysight. But sometimes the the format of the Thru standard is described in a format like this:
![ac2](https://user-images.githubusercontent.com/54358761/228245077-ee9ac917-523d-4084-93ef-7cf0c3888eda.png)
If you want the frequency dependent loss you can calculate it with the formula from this [paper ](https://cdn.teledynelecroy.com/files/appnotes/lecroy_sparq_cascade_microtech_s-parameter_measurements_rev1.pdf):
![loss](https://user-images.githubusercontent.com/54358761/228248340-b30d83af-2c6d-49a3-b6f4-6efe703e622d.png)
If iam using this formula for the offset_loss like this:
```python
freq = skrf.Frequency(100, 67000, 670, "MHz")

ref_loss = 0.34 
ref_delay = 25.5e-12
ref_freq = 20e9
thru_delay = 1.11e-12
offset_length_rs, offset_loss_rs =  keysight_to_rs(thru_delay, 1337, offset_z0=50)
thru_std = rs_calkit_thru(
    freq,
    offset_length=offset_length_rs, offset_loss=ref_loss*(thru_delay/ref_delay)/np.sqrt(ref_freq/1e9)
)

loss_thru = []
for f  in freq.f : loss_thru.append(ref_loss*(thru_delay/ref_delay)*np.sqrt(f/ref_freq)*-1)
```
Then my loss is lower than the loss if iam purely using the formula from the paper:
![output_1](https://user-images.githubusercontent.com/54358761/228250679-606d892a-8de4-4726-b5f7-126883b7b46f.png)
The factor is for some reason nearly 2. If i multiply the offset_loss with 2 iam getting the result of the formula:
```python
thru_std = rs_calkit_thru(
    freq,
    offset_length=offset_length_rs, offset_loss=ref_loss*(thru_delay/ref_delay)/np.sqrt(ref_freq/1e9)*2
)
```
![output_2](https://user-images.githubusercontent.com/54358761/228251657-d473c081-760e-4651-9180-1532cbf77afe.png)

Can anybody explain me why i have to multiply the offset_loss with 2 to achieve the results of the formula?

Here is the entire code:
```python
import skrf
from skrf.media import DefinedGammaZ0
import numpy as np
import matplotlib.pyplot as plt
skrf.stylely()

def rs_to_keysight(rs_offset_length, rs_offset_loss, offset_z0=50):
    offset_delay = rs_offset_length / skrf.constants.c
    offset_loss = skrf.mathFunctions.db_2_np(rs_offset_loss * offset_z0 / offset_delay)
    return offset_delay, offset_loss


def keysight_to_rs(keysight_offset_delay, keysight_offset_loss, offset_z0=50):
    offset_length = keysight_offset_delay * skrf.constants.c
    offset_loss = skrf.mathFunctions.np_2_db(keysight_offset_loss)/offset_z0 * keysight_offset_delay
    return offset_length, offset_loss


def rs_calkit_offset_line(freq, rs_offset_length, rs_offset_loss, offset_z0, port_z0):
    if rs_offset_length or rs_offset_loss:
        offset_delay, offset_loss = rs_to_keysight(rs_offset_length, rs_offset_loss)

        alpha_l = (offset_loss * offset_delay) / (2 * offset_z0)
        alpha_l *= np.sqrt(freq.f / 1e9)
        beta_l = 2 * np.pi * freq.f * offset_delay + alpha_l
        zc = offset_z0 + (1 - 1j) * (offset_loss / (4 * np.pi * freq.f)) * np.sqrt(freq.f / 1e9)
        gamma_l = alpha_l + beta_l * 1j

        medium = DefinedGammaZ0(frequency=freq, z0=port_z0, Z0=zc, gamma=gamma_l)
        offset_line = medium.line(d=1, unit='m', z0=medium.Z0, embed=True)
        return medium, offset_line
    else:
        medium = DefinedGammaZ0(frequency=freq, Z0=offset_z0)
        line = medium.line(d=0)
        return medium, line


def rs_calkit_open(freq, offset_length, offset_loss, c0, c1, c2, c3, offset_z0=50, port_z0=50):
    # Capacitance is defined with respect to the port impedance offset_z0, not the lossy
    # line impedance. In scikit-rf, the return values of `shunt_capacitor()` and `medium.open()`
    # methods are (correctly) referenced to the port impedance.
    medium, line = rs_calkit_offset_line(freq, offset_length, offset_loss, offset_z0, port_z0)
    if c0 or c1 or c2 or c3:
        poly = np.poly1d([c3, c2, c1, c0])
        capacitance = medium.shunt_capacitor(poly(freq.f)) ** medium.open()
    else:
        capacitance = medium.open()
    return line ** capacitance


def rs_calkit_short(freq, offset_length, offset_loss, l0, l1, l2, l3, offset_z0=50, port_z0=50):
    # Inductance is defined with respect to the port impedance offset_z0, not the lossy
    # line impedance. In scikit-rf, the return values of `inductor()` and `medium.short()`
    # methods are (correctly) referenced to the port impedance.
    medium, line = rs_calkit_offset_line(freq, offset_length, offset_loss, offset_z0, port_z0)
    if l0 or l1 or l2 or l3:
        poly = np.poly1d([l3, l2, l1, l0])
        inductance = medium.inductor(poly(freq.f)) ** medium.short()
    else:
        inductance = medium.short()
    return line ** inductance


def rs_calkit_load(freq, l0, offset_length=0, offset_loss=0, offset_z0=50, port_z0=50):
    medium, line = rs_calkit_offset_line(freq, offset_length, offset_loss, offset_z0, port_z0)
    if l0:
        poly = np.poly1d([0, 0, 0, l0])
        load = medium.inductor(poly(freq.f)) ** medium.match()
    else:
        load = medium.match()
    return line ** load


def rs_calkit_thru(freq, offset_length=0, offset_loss=0, offset_z0=50, port_z0=50):
    medium, line = rs_calkit_offset_line(freq, offset_length, offset_loss, offset_z0, port_z0)
    thru = medium.thru()
    return line ** thru


def keysight_calkit_offset_line(freq, offset_delay, offset_loss, offset_z0, port_z0):
    if offset_delay or offset_loss:
        alpha_l = (offset_loss * offset_delay) / (2 * offset_z0)
        alpha_l *= np.sqrt(freq.f / 1e9)
        beta_l = 2 * np.pi * freq.f * offset_delay + alpha_l
        zc = offset_z0 + (1 - 1j) * (offset_loss / (4 * np.pi * freq.f)) * np.sqrt(freq.f / 1e9)
        gamma_l = alpha_l + beta_l * 1j

        medium = DefinedGammaZ0(frequency=freq, z0=port_z0, Z0=zc, gamma=gamma_l)
        offset_line = medium.line(d=1, unit='m', z0=medium.Z0, embed=True)
        return medium, offset_line
    else:
        medium = DefinedGammaZ0(frequency=freq, Z0=offset_z0)
        line = medium.line(d=0)
        return medium, line


def keysight_calkit_open(freq, offset_delay, offset_loss, c0, c1, c2, c3, offset_z0=50, port_z0=50):
    medium, line = keysight_calkit_offset_line(freq, offset_delay, offset_loss, offset_z0, port_z0)
    # Capacitance is defined with respect to the port impedance offset_z0, not the lossy
    # line impedance. In scikit-rf, the return values of `shunt_capacitor()` and `medium.open()`
    # methods are (correctly) referenced to the port impedance.
    if c0 or c1 or c2 or c3:
        poly = np.poly1d([c3, c2, c1, c0])
        capacitance = medium.shunt_capacitor(poly(freq.f)) ** medium.open()
    else:
        capacitance = medium.open()
    return line ** capacitance


def keysight_calkit_short(freq, offset_delay, offset_loss, l0, l1, l2, l3, offset_z0=50, port_z0=50):
    # Inductance is defined with respect to the port impedance offset_z0, not the lossy
    # line impedance. In scikit-rf, the return values of `inductor()` and `medium.short()`
    # methods are (correctly) referenced to the port impedance.
    medium, line = keysight_calkit_offset_line(freq, offset_delay, offset_loss, offset_z0, port_z0)
    if l0 or l1 or l2 or l3:
        poly = np.poly1d([l3, l2, l1, l0])
        inductance = medium.inductor(poly(freq.f)) ** medium.short()
    else:
        inductance = medium.short()
    return line ** inductance


def keysight_calkit_load(freq, l0, offset_delay=0, offset_loss=0, offset_z0=50, port_z0=50):
    medium, line = keysight_calkit_offset_line(freq, offset_delay, offset_loss, offset_z0, port_z0)
    if l0:
        poly = np.poly1d([0, 0, 0, l0])
        load = medium.inductor(poly(freq.f)) ** medium.match()
    else:
        load = medium.match()
    return line ** load


def keysight_calkit_thru(freq, offset_delay=0, offset_loss=0, offset_z0=50, port_z0=50):
    medium, line = keysight_calkit_offset_line(freq, offset_delay, offset_loss, offset_z0, port_z0)
    thru = medium.thru()
    return line ** thru

freq = skrf.Frequency(100, 67000, 670, "MHz")

ref_loss = 0.34 
ref_delay = 25.5e-12
ref_freq = 20e9
thru_delay = 1.11e-12
offset_length_rs, offset_loss_rs =  keysight_to_rs(thru_delay, 1337, offset_z0=50)
thru_std = rs_calkit_thru(
    freq,
    offset_length=offset_length_rs, offset_loss=ref_loss*(thru_delay/ref_delay)/np.sqrt(ref_freq/1e9)*2
)

loss_thru = []
for f  in freq.f : loss_thru.append(ref_loss*(thru_delay/ref_delay)*np.sqrt(f/ref_freq)*-1)

PlotSize = 8
fig = plt.figure(figsize=(PlotSize, PlotSize))
ax = plt.subplot(111)
plt.plot(freq.f*1e-9, loss_thru,label="AC2-Formula",linewidth=2)
plt.plot(freq.f*1e-9, thru_std.s_db[:,1,0],label="AC2-RS",linewidth=2)
plt.legend()
plt.xlabel('Frequency (GHz)')
plt.ylabel('Mag (dB)')
box = ax.get_position()
ax.set_position([box.x0, box.y0 + box.height * 0.1,
                 box.width, box.height * 0.9])
plt.show()   
```
Thanks

**Comment by Vinc0110** [comment]
@biergaizi Can you help?

**Comment by biergaizi** [comment, marked as answer]
In Keysight and R&S's model, "offset loss" is defined to be a two-way loss from S11 reflection measurements. Thus, if the offset loss is obtained via one-way S21 measurements, it must be multiplied by two.

See *[Specifying Calibration Standards and Kits for Agilent Vector Network Analyzers](https://docplayer.net/storage/40/20954682/1680220059/__80gyyJyyU7oBLCCJNMfg/20954682.pdf)*, Equation 1.6, 1.7 for the definition of offset loss.

> * Offset loss
> Offset loss is in GΩ/sec. It is the propagation loss per unit length of the transmission line at a normalization frequency, such as 1 GHz, multiplied by the speed of light in the transmission medium. For coaxial devices, it can be calculated from the loss magnitude data at 1 GHz.
>
> S21 data (linear mag):
> ```math
>  \text{offset loss} = -2 \cdot \ln{(|S_{21}|}_\text{1 GHz}) \cdot \frac{\text{offset }Z_0}{\text{offset delay}}
> ```
> S11 data (linear mag):
> ```math
>  \text{offset loss} = - \ln{(|S_{11}|}_\text{1 GHz}) \cdot \frac{\text{offset }Z_0}{\text{offset delay}}
> ```

I believe both Keysight and R&S define offset loss this way, and it applies to both Reflection standards and Thru standards. If you believe this is incorrect or inconsistent with other references, software, or calculations, I'm open to correction. Otherwise, I think this is a satisfactory explanation of the behavior you are seeing here, to my best knowledge.

**Comment by Zwelckovich** [comment]
Thank you, 
thats the solution.
